In [1]:
# Imports and pip installations (if needed)
import pandas as pd

# Part 1: Load the dataset

In [2]:
# Load the given datasets
categorical_df = pd.read_csv("chronic_kidney_disease_categorical.csv")

# Print the data
categorical_df

,unique_id,al,su,rbc,pc,pcc,ba,htn,dm,cad,appet,pe,ane,Target
0,224481,3.0,NaN,NaN,NaN,notpresent,notpresent,yes,yes,no,good,yes,no,ckd
1,992643,0.0,0.0,NaN,NaN,notpresent,notpresent,no,no,no,good,no,no,notckd
2,308740,1.0,0.0,abnormal,abnormal,notpresent,notpresent,yes,no,no,good,yes,no,ckd
3,450314,1.0,0.0,NaN,normal,notpresent,notpresent,no,no,no,good,no,yes,ckd
4,881763,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
425,769512,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
426,872034,2.0,1.0,abnormal,abnormal,notpresent,present,yes,yes,yes,poor,no,no,ckd
427,643007,NaN,NaN,NaN,normal,notpresent,notpresent,yes,yes,yes,poor,no,no,ckd
428,840392,NaN,NaN,NaN,NaN,notpresent,notpresent,yes,no,no,poor,no,no,ckd


In [3]:
# Load the given datasets
numerical_df = pd.read_csv("chronic_kidney_disease_numerical.csv")

# Print the data
numerical_df

,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc
0,203694,38.0,80.0,99.0,19.0,0.5,147.0,3.5,13.6,44.0,7300.0,6.4
1,938027,43.0,60.0,108.0,25.0,1.0,144.0,5.0,17.8,43.0,7200.0,5.5
2,421471,37.0,60.0,111.0,35.0,0.8,135.0,4.1,16.2,50.0,5500.0,5.7
3,764115,70.0,90.0,144.0,125.0,4.0,136.0,4.6,12.0,37.0,8200.0,4.5
4,240975,47.0,80.0,95.0,35.0,0.9,140.0,4.1,NaN,NaN,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...
415,522912,56.0,80.0,165.0,55.0,1.8,NaN,NaN,13.5,40.0,11800.0,5.0
416,475200,64.0,100.0,163.0,54.0,7.2,140.0,4.6,7.9,26.0,7500.0,3.4
417,168242,48.0,60.0,112.0,44.0,1.2,142.0,4.9,14.5,44.0,9400.0,6.4
418,474407,21.0,90.0,107.0,40.0,1.7,125.0,3.5,8.3,23.0,12400.0,3.9


# Part 2: Analyze the Dataset

Refer to this: https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease

Explain what the each data is in your own words. What are the features and labels? Are the features in the given datasets : categorical, numerical or both? Give 3 examples of categorical and numerical columns each (if they exist)

Answer:
### Features and Labels
- **Features**: Used for prediction, including medical measurements and test results.
- **Label**: A variable indicating whether a patient has Chronic Kidney Disease (CKD).

### Features in the Datasets
The dataset contains both categorical and numerical features. 

#### Categorical Columns:
**rbc**, **htn**, **dm**

#### Numerical Columns:
**age**, **bp**, **bgr**

# Part 3: Data Preprocessing

A fundamental skill in Machine Learning is mastering the art of data cleaning and preprocessing. In this assignment, you will learn and apply essential data cleaning techniques to transform a raw dataset into a clean, ready-to-use form which you can use for regression or classification tasks. By the end of this assignment, you'll have a fully clean dataset and a solid foundation in preparing data for various machine learning models.

## Part 3.1 : Drop Duplicate rows

Let's start by checking if the given datasets have any duplicate rows (same Unique Id). Use pandas to identify and remove these duplicate rows from the given dataset

In [4]:
# For the numerical dataset, check if there are duplicate rows in the dataset. If yes, print total number of duplicate rows
num_duplicates = numerical_df.duplicated(subset = 'unique_id').sum()
print(f"Number of duplicate rows in numerical dataset: {num_duplicates}")

# Drop these duplicate rows
numerical_df = numerical_df.drop_duplicates(subset=['unique_id'])

# Repeat the same for categorical dataset. Print the duplicate rows and drop them
categorical_duplicates = categorical_df.duplicated(subset = 'unique_id').sum()
print(f"Number of duplicate rows in categorical dataset: {categorical_duplicates}")

# Drop these duplicate rows
categorical_df = categorical_df.drop_duplicates(subset=['unique_id'])


Number of duplicate rows in numerical dataset: 30
Number of duplicate rows in categorical dataset: 30


## Part 3.2: Combine two differents datasets

A good skill to have is to know how to combine 2 different datasets.

Are all the unique ids are present in both datasets? Why do you think so? If not, what do the rows that are missing from one of the datasets look like in the combined table?

Answer:

Not all unique ids may be present in both datasets due to differences in data collection methods. For example, different datasets may include different unique ids depending on who underwent certain tests.

When combining the two datasets, rows missing from one dataset will appear as NaN in the columns of that dataset in the combined table. This indicates that there is no corresponding data for those unique ids in the dataset they are missing from.

In [5]:
# Merge the numerical and categorical datasets based on the 'Unique Id' column
merged_df = pd.merge(categorical_df, numerical_df, on='unique_id')

# Print the merged dataframe
print(merged_df)

     unique_id   al   su       rbc        pc         pcc          ba  htn  \
0       224481  3.0  NaN       NaN       NaN  notpresent  notpresent  yes   
1       992643  0.0  0.0       NaN       NaN  notpresent  notpresent   no   
2       308740  1.0  0.0  abnormal  abnormal  notpresent  notpresent  yes   
3       450314  1.0  0.0       NaN    normal  notpresent  notpresent   no   
4       881763  0.0  0.0    normal    normal  notpresent  notpresent   no   
..         ...  ...  ...       ...       ...         ...         ...  ...   
385     789906  0.0  0.0    normal    normal  notpresent  notpresent   no   
386     490253  0.0  0.0    normal    normal  notpresent  notpresent   no   
387     492993  3.0  0.0       NaN    normal  notpresent  notpresent  yes   
388     273038  0.0  0.0    normal    normal  notpresent  notpresent   no   
389     583971  1.0  0.0       NaN    normal  notpresent  notpresent  yes   

      dm cad  ...    bp    bgr     bu   sc    sod  pot  hemo   pcv    wbcc 

## Part 3.3: Rows with Missing values

Removing missing values from a dataset is important for classification because it ensures the model is trained on complete and accurate data, leading to better performance and reliable predictions. Incomplete data can introduce bias and errors, negatively impacting the model's effectiveness.

In [6]:
# Calculate the percentage of rows that contain atleast one missing value
missing_percentage = (merged_df.isnull().any(axis=1).sum() / len(merged_df)) * 100

# Print %
print(f"Percentage of rows with at least one missing value in numerical dataset: {missing_percentage:.2f}%")

# Drop these rows from the dataset
merged_df = merged_df.dropna()

# Print the Dataset
print(merged_df)

Percentage of rows with at least one missing value in numerical dataset: 60.77%
     unique_id   al   su       rbc      pc         pcc          ba  htn   dm  \
4       881763  0.0  0.0    normal  normal  notpresent  notpresent   no   no   
5       114717  0.0  0.0    normal  normal  notpresent  notpresent   no   no   
7       901528  0.0  0.0    normal  normal  notpresent  notpresent   no   no   
8       433613  0.0  0.0    normal  normal  notpresent  notpresent   no   no   
9       621332  1.0  0.0  abnormal  normal  notpresent  notpresent   no  yes   
..         ...  ...  ...       ...     ...         ...         ...  ...  ...   
382     996076  0.0  0.0    normal  normal  notpresent  notpresent   no   no   
384     397388  2.0  2.0    normal  normal  notpresent     present  yes   no   
385     789906  0.0  0.0    normal  normal  notpresent  notpresent   no   no   
386     490253  0.0  0.0    normal  normal  notpresent  notpresent   no   no   
388     273038  0.0  0.0    normal  norm

## Part 3.4: Sort the dataset according to the Labels

In [7]:
# Sort the dataset according to the values in 'Target' column. Make sure reset the indices after sorting
merged_df_sorted = merged_df.sort_values(by='Target').reset_index(drop=True)

# Print the dataset
print(merged_df_sorted)

     unique_id   al   su       rbc        pc         pcc          ba  htn  \
0       534245  2.0  0.0    normal  abnormal     present  notpresent   no   
1       349892  1.0  3.0  abnormal  abnormal  notpresent  notpresent  yes   
2       343710  2.0  0.0  abnormal  abnormal  notpresent  notpresent  yes   
3       102677  3.0  0.0    normal  abnormal  notpresent  notpresent  yes   
4       514721  2.0  0.0  abnormal  abnormal  notpresent  notpresent  yes   
..         ...  ...  ...       ...       ...         ...         ...  ...   
148     489160  0.0  0.0    normal    normal  notpresent  notpresent   no   
149     530138  0.0  0.0    normal    normal  notpresent  notpresent   no   
150     261137  0.0  0.0    normal    normal  notpresent  notpresent   no   
151     322882  0.0  0.0    normal    normal  notpresent  notpresent   no   
152     273038  0.0  0.0    normal    normal  notpresent  notpresent   no   

      dm  cad  ...    bp    bgr     bu    sc    sod  pot  hemo   pcv     wb

## Part 3.5: Encoding Categorical data

In this step, we identify and process the categorical columns in the sorted dataset. We map each unique value in these columns to separate "dummy" columns.

For example, the column 'rbc' will be transformed into two columns 'rbc_normal' and 'rbc_abnormal'. If a row's value in 'rbc' is 'normal', the 'rbc_normal' column will be set to 1 and 'rbc_abnormal' will be set to 0.


**Note: Find a correct pandas function to do this **

In [8]:
# Write code here
categorical_columns = ['rbc', 'pc', 'pcc', 'ba', 'htn', 'dm', 'cad', 'appet', 'pe', 'ane', 'Target']

# Encoding categorical data using one-hot encoding
encoded_df = pd.get_dummies(merged_df_sorted, columns=categorical_columns, drop_first=True)

# Print the dataset
print(encoded_df)

     unique_id   al   su   age    bp    bgr     bu    sc    sod  pot  ...  \
0       534245  2.0  0.0  45.0  70.0  117.0   52.0   2.2  136.0  3.8  ...   
1       349892  1.0  3.0  59.0  70.0  424.0   55.0   1.7  138.0  4.5  ...   
2       343710  2.0  0.0  61.0  80.0  173.0  148.0   3.9  135.0  5.2  ...   
3       102677  3.0  0.0  59.0  70.0   76.0  186.0  15.0  135.0  7.6  ...   
4       514721  2.0  0.0  56.0  90.0  129.0  107.0   6.7  131.0  4.8  ...   
..         ...  ...  ...   ...   ...    ...    ...   ...    ...  ...  ...   
148     489160  0.0  0.0  28.0  60.0   79.0   50.0   0.5  145.0  5.0  ...   
149     530138  0.0  0.0  23.0  80.0  111.0   34.0   1.1  145.0  4.0  ...   
150     261137  0.0  0.0  35.0  60.0  105.0   39.0   0.5  135.0  3.9  ...   
151     322882  0.0  0.0  33.0  80.0  128.0   38.0   0.6  135.0  3.9  ...   
152     273038  0.0  0.0  47.0  60.0  117.0   22.0   1.2  138.0  3.5  ...   

     pc_normal  pcc_present  ba_present  htn_yes  dm_yes  cad_yes  appet_po

In the example we went through above, another solution is to have a single column for the binary variable. In the downstream modeling would this be equivalent? What effect would this have if the categorical variable could take more than 2 values? For example, let's say we have a categorical feature that is "type of condiment" that can take 5 separate values and we are trying to predict the rating of a particular sandwich.

## Part 3.6 : Remove Outliers from Numerical Columns

Outliers can disproportionately influence the fit of a regression model, potentially leading to a model that does not generalize well therefore it is important that we remove outliers from the numerical columns of the dataset.

For this dataset, we define an outlier to be 3 times the standard deviation from the mean. Drop these outliers from the dataset

In [9]:
# Remove outliers
numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']

# Define the threshold for outliers (3 standard deviations)
threshold = 3

# Loop through the numerical columns and filter out the outliers
for col in numerical_columns:
    mean = encoded_df[col].mean()
    std = encoded_df[col].std()
    
    # Keep only the rows where the values are within 3 standard deviations
    encoded_df = encoded_df[(encoded_df[col] >= mean - threshold * std) & (encoded_df[col] <= mean + threshold * std)]

# Print the dataset
print(encoded_df)

     unique_id   al   su   age    bp    bgr     bu   sc    sod  pot  ...  \
2       343710  2.0  0.0  61.0  80.0  173.0  148.0  3.9  135.0  5.2  ...   
4       514721  2.0  0.0  56.0  90.0  129.0  107.0  6.7  131.0  4.8  ...   
7       124304  3.0  4.0  69.0  70.0  264.0   87.0  2.7  130.0  4.0  ...   
8       980291  3.0  0.0  62.0  70.0  122.0   42.0  1.7  136.0  4.7  ...   
10      615697  4.0  0.0   6.0  60.0   94.0   67.0  1.0  135.0  4.9  ...   
..         ...  ...  ...   ...   ...    ...    ...  ...    ...  ...  ...   
148     489160  0.0  0.0  28.0  60.0   79.0   50.0  0.5  145.0  5.0  ...   
149     530138  0.0  0.0  23.0  80.0  111.0   34.0  1.1  145.0  4.0  ...   
150     261137  0.0  0.0  35.0  60.0  105.0   39.0  0.5  135.0  3.9  ...   
151     322882  0.0  0.0  33.0  80.0  128.0   38.0  0.6  135.0  3.9  ...   
152     273038  0.0  0.0  47.0  60.0  117.0   22.0  1.2  138.0  3.5  ...   

     pc_normal  pcc_present  ba_present  htn_yes  dm_yes  cad_yes  appet_poor  \
2     

## Part 3.7 : Normalize the Numerical Columns

Normalizing numerical attributes ensures that all features contribute equally to the model by scaling them to a consistent range, which improves model performance and convergence. It prevents features with larger scales from disproportionately influencing the model's learning process.

In [10]:
# Normalize the all Numerical Attributes in the dataset.
for column in numerical_columns:
    min_value = encoded_df[column].min()
    max_value = encoded_df[column].max()
    
    # Apply min-max scaling
    encoded_df[column] = (encoded_df[column] - min_value) / (max_value - min_value)

# Print the dataset
print(encoded_df)

     unique_id   al   su       age    bp       bgr        bu        sc  \
2       343710  2.0  0.0  0.743243  0.50  0.442060  0.901961  0.432099   
4       514721  2.0  0.0  0.675676  0.75  0.253219  0.633987  0.777778   
7       124304  3.0  4.0  0.851351  0.25  0.832618  0.503268  0.283951   
8       980291  3.0  0.0  0.756757  0.25  0.223176  0.209150  0.160494   
10      615697  4.0  0.0  0.000000  0.00  0.103004  0.372549  0.074074   
..         ...  ...  ...       ...   ...       ...       ...       ...   
148     489160  0.0  0.0  0.297297  0.00  0.038627  0.261438  0.012346   
149     530138  0.0  0.0  0.229730  0.50  0.175966  0.156863  0.086420   
150     261137  0.0  0.0  0.391892  0.00  0.150215  0.189542  0.012346   
151     322882  0.0  0.0  0.364865  0.50  0.248927  0.183007  0.024691   
152     273038  0.0  0.0  0.554054  0.00  0.201717  0.078431  0.098765   

          sod       pot  ...  pc_normal  pcc_present  ba_present  htn_yes  \
2    0.500000  0.793103  ...      

## Part 3.8: Remove Unnecessary columns

Are there any columns in this dataset which are not appropriate for modeling and predictions? Which column(s)? Justify their exclusion and remove them

Answer:

The "unique_id" column is simply an identifier for each patient and does not contain any medical or diagnostic information that would help in predicting CKD. It has no correlation with the target variable and would not provide any value to the model's learning process. Including it could potentially introduce noise.

In [11]:
#Remove that column
encoded_df = encoded_df.drop('unique_id', axis=1)

# Print the dataset
print(encoded_df)

      al   su       age    bp       bgr        bu        sc       sod  \
2    2.0  0.0  0.743243  0.50  0.442060  0.901961  0.432099  0.500000   
4    2.0  0.0  0.675676  0.75  0.253219  0.633987  0.777778  0.366667   
7    3.0  4.0  0.851351  0.25  0.832618  0.503268  0.283951  0.333333   
8    3.0  0.0  0.756757  0.25  0.223176  0.209150  0.160494  0.533333   
10   4.0  0.0  0.000000  0.00  0.103004  0.372549  0.074074  0.500000   
..   ...  ...       ...   ...       ...       ...       ...       ...   
148  0.0  0.0  0.297297  0.00  0.038627  0.261438  0.012346  0.833333   
149  0.0  0.0  0.229730  0.50  0.175966  0.156863  0.086420  0.833333   
150  0.0  0.0  0.391892  0.00  0.150215  0.189542  0.012346  0.500000   
151  0.0  0.0  0.364865  0.50  0.248927  0.183007  0.024691  0.500000   
152  0.0  0.0  0.554054  0.00  0.201717  0.078431  0.098765  0.600000   

          pot      hemo  ...  pc_normal  pcc_present  ba_present  htn_yes  \
2    0.793103  0.000000  ...      False       

## Part 3.9: Export the Cleaned Data

Now that you've completed these cleaning steps you should have a pandas dataframe which is much cleaner and ready for modeling. Our final step is to save our work. Export the DataFrame to a two new formats: csv and json.

In [12]:
# Export the dataframe to a new csv file
encoded_df.to_csv("chronic_kidney_disease_cleaned.csv", index = False)

# Export the dataframe to a new json file
encoded_df.to_json("chronic_kidney_disease_cleaned.json")

# Part 4: Data conversions with Large Language Models

One powerful use case of ChatGPT (and other generative language models) is cleaning and transforming data. In some cases, these models can directly manipulate loosely structured data that you provide to them into a standard format. In the other cases, you can often prompt the model to create a conversion or extraction script for you in python or Pandas and then run it on your own. 

In this part of the assignment you will prompt 383GPT to explore these capabilities.

## Part 4.1 GPT Data Manipulation

Take the cleaned dataset that you created in part three and output the top 15 rows of that dataset. Then copy the terminal output, open 383gpt and ask it to convert that output to a markdown table. Paste that markdown table in the cell bellow

In [13]:
chronic_kidney_disease_cleaned_15_df = pd.read_csv("chronic_kidney_disease_cleaned.csv")
chronic_kidney_disease_cleaned_15_df.head(15)

,al,su,age,bp,bgr,bu,sc,sod,pot,hemo,...,pc_normal,pcc_present,ba_present,htn_yes,dm_yes,cad_yes,appet_poor,pe_yes,ane_yes,Target_notckd
0,2.0,0.0,0.743243,0.50,0.442060,0.901961,0.432099,0.500000,0.793103,0.000000,...,False,False,False,True,True,True,True,True,True,False
1,2.0,0.0,0.675676,0.75,0.253219,0.633987,0.777778,0.366667,0.655172,0.138614,...,False,False,False,True,False,False,False,False,False,False
2,3.0,4.0,0.851351,0.25,0.832618,0.503268,0.283951,0.333333,0.379310,0.475248,...,False,False,False,True,True,True,False,True,False,False
3,3.0,0.0,0.756757,0.25,0.223176,0.209150,0.160494,0.533333,0.620690,0.485149,...,False,False,False,True,True,False,False,False,False,False
4,4.0,0.0,0.000000,0.00,0.103004,0.372549,0.074074,0.500000,0.689655,0.217822,...,False,False,True,False,False,False,True,False,False,False
5,3.0,1.0,0.662162,0.50,0.618026,0.411765,0.432099,0.566667,0.689655,0.316832,...,False,True,True,True,True,False,False,True,False,False
6,1.0,0.0,0.540541,0.00,0.399142,0.535948,0.358025,0.700000,0.379310,0.207921,...,True,False,False,True,True,False,False,False,False,False
7,4.0,2.0,0.783784,1.00,0.399142,0.287582,0.839506,0.666667,0.586207,0.019802,...,False,False,True,True,True,False,False,True,False,False
8,4.0,0.0,0.567568,0.50,0.107296,1.000000,0.901235,0.533333,0.310345,0.207921,...,False,False,False,True,False,False,False,False,True,False
9,4.0,3.0,0.851351,0.25,0.618026,0.562092,0.728395,0.000000,0.344828,0.168317,...,False,True,True,True,True,True,False,True,True,False


### Paste the markdown table here

| al  | su  | age                 | bp   | bgr                 | bu                  | sc                  | sod                 | pot                 | hemo  | pcv                | wbcc                | rbcc                | rbc_normal | pc_normal | pcc_present | ba_present | htn_yes | dm_yes | cad_yes | appet_poor | pe_yes | ane_yes | Target_notckd |
|-----|-----|---------------------|------|---------------------|---------------------|---------------------|---------------------|---------------------|-------|---------------------|---------------------|---------------------|------------|-----------|-------------|------------|---------|--------|---------|-------------|--------|---------|----------------|
| 2.0 | 0.0 | 0.7432432432432432  | 0.5  | 0.44206008583690987 | 0.9019607843137255  | 0.4320987654320988  | 0.5                 | 0.7931034482758622  | 0.0   | 0.03225806451612903 | 0.3951612903225806  | 0.0571428571428572  | False      | False     | False       | False      | True    | True   | True    | True        | True   | True    | False          |
| 2.0 | 0.0 | 0.6756756756756757  | 0.75 | 0.2532188841201717  | 0.6339869281045751  | 0.7777777777777778  | 0.36666666666666664 | 0.6551724137931034  | 0.1386| 0.1935483870967742  | 0.1693548387096774  | 0.11428571428571425 | False      | False     | False       | False      | True    | False  | False   | False       | False   | False    | False          |
| 3.0 | 4.0 | 0.8513513513513513  | 0.25 | 0.8326180257510729  | 0.5032679738562091  | 0.28395061728395066 | 0.3333333333333333  | 0.37931034482758624 | 0.475 | 0.45161290322580644 | 0.4274193548387097  | 0.31428571428571417 | True       | False     | False       | False      | True    | True   | True    | False       | True    | False    | False          |
| 3.0 | 0.0 | 0.7567567567567568  | 0.25 | 0.22317596566523606 | 0.20915032679738563 | 0.16049382716049382 | 0.5333333333333333  | 0.6206896551724139  | 0.485 | 0.5161290322580645  | 0.2903225806451613  | 0.2571428571428571  | True       | False     | False       | False      | True    | True   | False   | False       | False   | False    | False          |
| 4.0 | 0.0 | 0.0                 | 0.0  | 0.10300429184549356 | 0.37254901960784315 | 0.07407407407407407 | 0.5                 | 0.6896551724137933  | 0.2178| 0.22580645161290322 | 1.0                 | 0.5142857142857142  | False      | False     | False       | True       | False   | False  | False   | True        | False   | False    | False          |
| 3.0 | 1.0 | 0.6621621621621622  | 0.5  | 0.6180257510729614  | 0.4117647058823529  | 0.4320987654320988  | 0.5666666666666667  | 0.6896551724137933  | 0.3168| 0.3548387096774194  | 0.25                | 0.20000000000000004 | True       | False     | True        | True       | True    | True   | False   | False       | True    | False    | False          |
| 1.0 | 0.0 | 0.5405405405405406  | 0.0  | 0.39914163090128757 | 0.5359477124183006  | 0.35802469135802467 | 0.7                 | 0.37931034482758624 | 0.2079| 0.16129032258064516 | 0.8306451612903226  | 0.0571428571428572  | True       | True      | False       | False      | True    | True   | False   | False       | False   | False    | False          |
| 4.0 | 2.0 | 0.7837837837837838  | 1.0  | 0.39914163090128757 | 0.2875816993464052  | 0.8395061728395062  | 0.6666666666666666  | 0.5862| 0.019801980198019816 | 0.0967741935483871  | 0.25806451612903225 | 0.11428571428571425 | False      | False     | False       | True       | True    | True   | False   | False       | True    | False    | False          |
| 4.0 | 0.0 | 0.5675675675675675  | 0.5  | 0.1072961373390558  | 1.0                 | 0.9012345679012346  | 0.5333333333333333  | 0.3103| 0.20792079207920794 | 0.2903225806451613  | 0.20967741935483872 | 0.11428571428571425 | True       | False     | False       | False      | True    | False  | False   | False       | True    | False    | False          |
| 4.0 | 3.0 | 0.8513513513513513  | 0.25 | 0.6180257510729614  | 0.5620915032679739  | 0.7283950617283951  | 0.0                 | 0.3448| 0.1683168316831683  | 0.16129032258064516 | 0.5806451612903226  | 0.08571428571428566 | True       | False     | True        | True       | True    | True   | False   | True        | True    | False    | False          |
| 3.0 | 2.0 | 0.9054054054054054  | 1.0  | 0.9656652360515021  | 0.5228758169934641  | 0.6419753086419753  | 0.6666666666666666  | 0.0   | 0.1485148514851484  | 0.22580645161290322 | 0.21774193548387097 | 0.0571428571428572 | False      | False     | True        | False      | True    | True   | True    | True        | False   | False    | False          |
| 4.0 | 0.0 | 0.20270270270270271 | 0.75 | 0.15879828326180256 | 0.19607843137254902 | 0.16049382716049382 | 0.16666666666666666 | 0.2069| 0.05940594059405945 | 0.0                 | 0.6532258064516129  | 0.2571428571428571  | True       | False     | True        | True       | False   | False  | False   | False       | False   | True     | False          |
| 4.0 | 1.0 | 0.7837837837837838  | 0.0  | 0.7253218884120172  | 0.3137254901960784  | 0.4814814814814815  | 0.5666666666666667  | 0.8621| 0.17821782178217818 | 0.1935483870967742  | 0.25806451612903225 | 0.11428571428571425 | False      | False     | False       | True       | True    | True   | False   | True        | False   | False    | False          |
| 4.0 | 1.0 | 0.6756756756756757  | 0.25 | 0.6008583690987125  | 0.10457516339869281 | 0.16049382716049382 | 0.5333333333333333  | 0.3103| 0.8316831683168318  | 0.9354838709677419  | 0.6612903225806451  | 0.7428571428571428  | False      | True      | False       | False      | False   | False  | False   | False       | False   | False    | False          |
| 4.0 | 0.0 | 0.5675675675675675  | 0.5  | 0.2703862660944206  | 0.8431372549019608  | 1.0                 | 0.4                 | 0.8966| 0.25742574257425743 | 0.41935483870967744 | 0.1532258064516129  | 0.2857142857142857  | False      | False     | False       | True       | False   | True   | False   | False       | True    | False    | False          |


**Caution:** while language models can perform data conversions they also can *hallucinate* during this process, particularly for bigger datasets. Reflect on this below, how could you mitigate data conversion hallucinations from LLM conversions?

Answer:

To mitigate data conversion hallucinations from language model conversions, it's essential to:

1. **Verify Outputs**: Cross-check the converted data against the original dataset to ensure accuracy.
2. **Use Smaller Batches**: Break down larger datasets into smaller chunks for conversion, reducing the risk of errors.
3. **Provide Context**: Include clear instructions and context in the prompt to guide the model in producing the correct format.
4. **Implement Manual Checks**: Perform manual checks of the data for consistency and completeness.

## Part 4.2 GPT Pandas Prompting

In this section, you will prompt 383GPT to write pandas code manipulations for you.

After working with this data for awhile, we realized we're starting to forget the meanings of the abbreviated column names. Let's ask 383GPT to fix this for us. First, navigate to the [UCI dataset overview](https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease) and copy the abbrevation to name mapping. Then, go to 383GPT and instruct the LLM to provide you with a pandas script to apply this renaming to all the columns of your dataset. Paste that code below and make any adjustments necessary to run it in your notebook.

In [14]:
# Code to rename all the columns in the dataset

# Create a dictionary mapping old column names to new column names
rename_mapping = {
    'age': 'age',
    'bp': 'blood pressure',
    'sg': 'specific gravity',
    'al': 'albumin',
    'su': 'sugar',
    'rbc': 'red blood cells',
    'pc': 'pus cell',
    'pcc': 'pus cell clumps',
    'ba': 'bacteria',
    'bgr': 'blood glucose random',
    'bu': 'blood urea',
    'sc': 'serum creatinine',
    'sod': 'sodium',
    'pot': 'potassium',
    'hemo': 'hemoglobin',
    'pcv': 'packed cell volume',
    'wc': 'white blood cell count',
    'rc': 'red blood cell count',
    'htn': 'hypertension',
    'dm': 'diabetes mellitus',
    'cad': 'coronary artery disease',
    'appet': 'appetite',
    'pe': 'pedal edema',
    'ane': 'anemia',
    'class': 'class'
}

# Rename the columns of the DataFrame
chronic_kidney_disease_cleaned_15_df.rename(columns=rename_mapping, inplace=True)

# Display the updated DataFrame to confirm changes
chronic_kidney_disease_cleaned_15_df.head()


,albumin,sugar,age,blood pressure,blood glucose random,blood urea,serum creatinine,sodium,potassium,hemoglobin,...,pc_normal,pcc_present,ba_present,htn_yes,dm_yes,cad_yes,appet_poor,pe_yes,ane_yes,Target_notckd
0,2.0,0.0,0.743243,0.50,0.442060,0.901961,0.432099,0.500000,0.793103,0.000000,...,False,False,False,True,True,True,True,True,True,False
1,2.0,0.0,0.675676,0.75,0.253219,0.633987,0.777778,0.366667,0.655172,0.138614,...,False,False,False,True,False,False,False,False,False,False
2,3.0,4.0,0.851351,0.25,0.832618,0.503268,0.283951,0.333333,0.379310,0.475248,...,False,False,False,True,True,True,False,True,False,False
3,3.0,0.0,0.756757,0.25,0.223176,0.209150,0.160494,0.533333,0.620690,0.485149,...,False,False,False,True,True,False,False,False,False,False
4,4.0,0.0,0.000000,0.00,0.103004,0.372549,0.074074,0.500000,0.689655,0.217822,...,False,False,True,False,False,False,True,False,False,False


## Part 4.3 Augmenting our skills with prompting

In addition, we can also use 383GPT to convert our data manipulation operations between different data manipulation languages and libraries. For example let's prompt 383GPT to convert the following SQL query to a pandas query.

**SQL Query**
```sql
SELECT Target, COUNT(*) AS count
FROM your_table_name
GROUP BY Target;
```

Prompt 383GPT to convert this to a pandas query. Run this query below, then describe what it does. (If you're not familiar with SQL that is okay you need to only comment on the final resulting output.)

In [15]:
# Converted SQL to Pandas code
target_count = chronic_kidney_disease_cleaned_15_df.groupby('Target_notckd').size().reset_index(name='count')

# Display the resulting DataFrame
target_count

,Target_notckd,count
0,False,22
1,True,111


*Describe what the above code does here*

This code prints how many patients in the dataset have each category of CKD based on the values in the Target column, making it easier to understand the distribution of CKD in the dataset.